### PR TITLE
[Kernels]Updated Triton kernels into 2.1.0 and adding flash-decoding for llama token attention 

### PR DESCRIPTION
**Pull request by tiandiao123**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
This PR is used to add flash-decoding and latest triton version (2.1.0 )'s kernels into our inference system. 
Here is updated inference result using colossal-ai original kv-cache manager + triton kernels
'''
[10/27/23 10:28:25] INFO     colossalai - colossalai.inference.tensor_parallel.kvcache_manager - INFO: freed all space of memory manager          
generation time 6.005609035491943 s
128
[10/27/23 10:28:31] INFO     colossalai - colossalai.inference.tensor_parallel.kvcache_manager - INFO: freed all space of memory manager          
generation time 6.332996368408203 s
128
total process latency is : 0.04655854856329305 s
total throughput is : 687.3066491000737
'''

The following is flash-decoding result using kv-cache manager + flash-decoding:

'''
[10/27/23 10:21:57] INFO     colossalai - colossalai.inference.tensor_parallel.kvcache_manager - INFO: freed all space of memory manager          
generation time 8.618249654769897 s
128
[10/27/23 10:22:06] INFO     colossalai - colossalai.inference.tensor_parallel.kvcache_manager - INFO: freed all space of memory manager
128
total process latency is : 0.04120702562587602 s
total throughput is : 776.5666051835964

'''

both of them uses llama models, and set input-length as 1024 abd output as 128

Suggest separate current flash-decoding kernel and kv-cache manager. 

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/tensor_parallel/modeling/llama.py [hunk 1]
```diff
@@ -348,24 +358,34 @@ def llama_flash_attn_kvcache_forward(
...
+                    attn_output = torch.empty_like(query_states)        
+                    heads_per_group = self.num_heads // self.num_key_value_heads
+                    cache_k = infer_state.cache_manager.key_buffer[infer_state.decode_layer_id]
+                    cache_v = infer_state.cache_manager.value_buffer[infer_state.decode_layer_id]
+                    
+                    query_states = query_states.view(bsz, -1, self.num_heads, self.head_dim)
+                    cache_k = cache_k.view(bsz, -1, self.num_key_value_heads, self.head_dim)
+                    cache_v = cache_v.view(bsz, -1, self.num_key_value_heads, self.head_dim)
+                    
+                    attn_output = flash_attn_with_kvcache(q = query_states, k_cache = cache_k, v_cache = cache_v, softmax_scale = 1/ math.sqrt(self.head_dim), causal = True)
```
Comment (yuanheng-zhao) [inline colossalai/inference/tensor_parallel/modeling/llama.py:373]: I'm afraid that `flash_attn_with_kvcache` will mix up the kv cache we have at this moment. For now, k/v cache is assigned on a block of cache space and managed by cache manager via block location indices, start indices, and new kv will be added into kv cache via kernel function like `copy_kv_toXX`. However, `flash_attn_with_kvcache` will update the passed `k_cache` and `v_cache` in-place, which means it interferes the indexing system and manipulates the cache space outside previous ways.

Comment (tiandiao123) [reply colossalai/inference/tensor_parallel/modeling/llama.py:372]: OK！ probably need to write a new forward function?

File: colossalai/inference/tensor_parallel/modeling/llama.py [hunk 2]
```diff
@@ -348,35 +385,27 @@ def llama_flash_attn_kvcache_forward(
...
-                )
+                heads_per_group = self.num_heads // self.num_key_value_heads
+                cache_k = infer_state.cache_manager.key_buffer[infer_state.decode_layer_id]
+                cache_v = infer_state.cache_manager.value_buffer[infer_state.decode_layer_id]
+                
+                query_states = query_states.view(bsz, -1, self.num_heads, self.head_dim)
+                copy_cache_k= copy.deepcopy(cache_k)
+                copy_cache_v = copy.deepcopy(cache_v)
+                copy_cache_k = copy_cache_k.view(bsz, -1, self.num_key_value_heads, self.head_dim)
+                copy_cache_v = copy_cache_v.view(bsz, -1, self.num_key_value_heads, self.head_dim)
```
Comment (CjhHa1) [inline colossalai/inference/tensor_parallel/modeling/llama.py:401]: `copied_kv` need to be delete after flash decoding?

**Comment by Xu-Kai** [comment]
Should we add lightllm in requirements-test for lightllm kernels?

File: colossalai/inference/README.md [hunk 3]
```diff
@@ -34,11 +34,13 @@ In this section we discuss how the colossal inference works and integrates with
   - [x] policy
   - [x] context forward
   - [x] token forward
+  - [] support flash-decoding
```
[the remark this answers is not in the document]
Comment (tiandiao123) [reply colossalai/inference/README.md]: fixed.

File: colossalai/inference/tensor_parallel/modeling/llama.py (hunk 2 above)
Comment (tiandiao123) [reply colossalai/inference/tensor_parallel/modeling/llama.py]: fixed.

**Comment by tiandiao123** [comment]
> Should we add lightllm in requirements-test for lightllm kernels?

we don't have pip install lightllm, we can only build from source

**Review by Xu-Kai — APPROVED** [review]

**Review by CjhHa1 — APPROVED** [review]
LGTM

**Pull request merged**